java/fury-core/src/main/java/org/apache/fury/memory/MemoryBuffer.java

```diff
@@ -100,24 +105,30 @@ public final class MemoryBuffer {
    * @param length buffer size
    */
   private MemoryBuffer(byte[] buffer, int offset, int length) {
-    Preconditions.checkArgument(offset >= 0 && length >= 0);
+    this(buffer, offset, length, null);
+  }
+
+  /**
+   * Creates a new memory buffer that represents the memory of the byte array.
+   *
+   * @param buffer The byte array whose memory is represented by this memory buffer.
+   * @param offset The offset of the sub array to be used; must be non-negative and no larger than
+   *     <tt>array.length</tt>.
+   * @param length buffer size
+   * @param streamReader a reader for reading from a stream.
+   */
+  private MemoryBuffer(byte[] buffer, int offset, int length, FuryStreamReader streamReader) {
+    checkArgument(offset >= 0 && length >= 0);
     if (offset + length > buffer.length) {
       throw new IllegalArgumentException(
           String.format("%d exceeds buffer size %d", offset + length, buffer.length));
     }
     initHeapBuffer(buffer, offset, length);
-  }
-
-  private void initHeapBuffer(byte[] buffer, int offset, int length) {
-    if (buffer == null) {
-      throw new NullPointerException("buffer");
+    if (streamReader != null) {
+      this.streamReader = streamReader;
+    } else {
+      this.streamReader = new BoundChecker();
```

Comment:
Proposal: maybe instead of `BoundChecker` class we can make `MemoryBuffer` class itself extending `AbstractStreamReader`? In this case we can avoid new class initialization by simply passing `this` to the `streamReader` field.

Reply:
Could you elaborate more how `MemoryBuffer` read from the `InputStream`/`Channel`? There still a interface for such cases.

Reply:
Note: we can't make `MemoryBuffer` as a polymorphic class, it will make a large degeneration to performance

Reply:
Yeah, sound reasonable. I just trying to find a way to avoid unnecessary object allocation, but can't find any better solution...

Reply:
I tried make boundchecker as a singleton, but then is can't access the index in memory buffer. Since memory buffer won't be created frequently. The object creation cost should be acceptable